[WK2] Add initial implementation the Screen Wake Lock API behind an experimental flag

cdumez opened the pull request:
#### 95606b13126e82a5cf2534a05e107d0aa69ccc0a
<pre>
[WK2] Add initial implementation the Screen Wake Lock API behind an experimental flag
<a href="https://bugs.webkit.org/show_bug.cgi?id=245712">https://bugs.webkit.org/show_bug.cgi?id=245712</a>

Reviewed by Alex Christensen, Brent Fulgham and Geoffrey Garen.

Add initial implementation the Screen Wake Lock API for WebKit2, behind an experimental flag:
- <a href="https://w3c.github.io/screen-wake-lock">https://w3c.github.io/screen-wake-lock</a>

The implementation should be complete but it is currently behind an experimental feature flag
that is off by default.

Some tests are still failing. The reason for that seems to be that our support for Feature
Policy is only partial. In particular, we seem to support the `allow` attribute on iframes
but not the Feature-Policy HTTP header.

We also have to make a decision about whether or not we want to prompt the user when the
Website tries to take a screen wake lock and if we need an indicator in the browser UI.
Currently, my implementation only requires a user gesture.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/feature-policy/resources/feature-policy-screen-wakelock.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/feature-policy/resources/featurepolicy.js:
(expect_reports):
* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
(window.test_driver_internal.set_permission):
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/idlharness.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-active-document.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-disabled-by-feature-policy.https.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-disabled-by-feature-policy.https.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-by-feature-policy-attribute-redirect-on-load.https.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-by-feature-policy-attribute-redirect-on-load.https.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-by-feature-policy-attribute.https.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-by-feature-policy-attribute.https.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-by-feature-policy.https.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-by-feature-policy.https.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-on-self-origin-by-feature-policy.https.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-on-self-origin-by-feature-policy.https.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-onrelease.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-released.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-request-denied.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-type.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelockpermissiondescriptor.https-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/Modules/permissions/PermissionName.h:
* Source/WebCore/Modules/permissions/PermissionName.idl:
* Source/WebCore/Modules/screen-wake-lock/NavigatorScreenWakeLock.cpp:
(WebCore::NavigatorScreenWakeLock::NavigatorScreenWakeLock):
(WebCore::NavigatorScreenWakeLock::wakeLock):
* Source/WebCore/Modules/screen-wake-lock/NavigatorScreenWakeLock.h:
* Source/WebCore/Modules/screen-wake-lock/WakeLock.cpp:
(WebCore::WakeLock::WakeLock):
(WebCore::WakeLock::request):
(WebCore::WakeLock::document):
* Source/WebCore/Modules/screen-wake-lock/WakeLock.h:
(WebCore::WakeLock::create):
* Source/WebCore/Modules/screen-wake-lock/WakeLockManager.cpp: Added.
(WebCore::WakeLockManager::WakeLockManager):
(WebCore::WakeLockManager::~WakeLockManager):
(WebCore::WakeLockManager::addWakeLock):
(WebCore::WakeLockManager::removeWakeLock):
(WebCore::WakeLockManager::visibilityStateChanged):
(WebCore::WakeLockManager::releaseAllLocks):
* Source/WebCore/Modules/screen-wake-lock/WakeLockManager.h: Copied from Source/WebCore/Modules/screen-wake-lock/WakeLock.h.
* Source/WebCore/Modules/screen-wake-lock/WakeLockSentinel.cpp:
(WebCore::WakeLockSentinel::release):
(WebCore::WakeLockSentinel::virtualHasPendingActivity const):
(WebCore::WakeLockSentinel::eventListenersDidChange):
(WebCore::WakeLockSentinel::wakeLockManager):
* Source/WebCore/Modules/screen-wake-lock/WakeLockSentinel.h:
* Source/WebCore/Modules/screen-wake-lock/WakeLockType.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::wakeLockManager):
(WebCore::Document::stopActiveDOMObjects):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/TaskSource.h:
* Source/WebCore/html/FeaturePolicy.cpp:
(WebCore::policyTypeName):
(WebCore::FeaturePolicy::parse):
(WebCore::FeaturePolicy::allows const):
* Source/WebCore/html/FeaturePolicy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::queryPermission):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::setScreenWakeLockPermission):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.h:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setScreenWakeLockPermission):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::handleQueryPermission):
(WTR::TestController::resetStateToConsistentValues):
(WTR::TestController::setScreenWakeLockPermission):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):

Canonical link: <a href="https://commits.webkit.org/255019@main">https://commits.webkit.org/255019@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/575238afbfbefdcc52d73b139fb44df509a85be0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90990 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35561 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21524 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/100336 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/158837 "Reverted pull request changes (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94996 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/34061 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83334 "Built successfully") | [💥 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97124 "An unexpected error occured. Recent messages:Failed to print configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96646 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77773 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26947 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81890 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/81614 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69982 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35140 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15635 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32938 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16622 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3496 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36718 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/39583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38646 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35706 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

marcoscaceres commented on Source/WebCore/Modules/screen-wake-lock/WakeLock.cpp:
```diff
@@ -26,16 +26,69 @@
+    }
+
+    // We should be requesting permission but at the moment, we merely require a user gesture.
+    bool hasUserGesture = UserGestureIndicator::processingUserGesture(document);
```
Should we use/consume transient activation instead?

cdumez replied:
I updated the patch to *use* transient activation. Let me know if you think we should *consume* instead.

geoffreygaren replied:
I would favor not consuming in this case; but we can have that discussion in the standards github. (I've commented there already.)

geoffreygaren commented on Source/WebCore/Modules/screen-wake-lock/WakeLockManager.cpp:
```diff
@@ -0,0 +1,105 @@
+
+WakeLockManager::~WakeLockManager()
+{
+    m_document.unregisterForVisibilityStateChangedCallbacks(*this);
```
I wish we could use a weak map for this instead.

cdumez replied:
Not sure what you mean. It is already a WeakHashSet:
`WeakHashSet<VisibilityChangeClient> m_visibilityStateCallbackClients;`

This doesn't mean we shouldn't remove the client when it is gone as far as I know. Our WeakHashSet implementation doesn't aggressively prune dead entries.

geoffreygaren replied:
I see: We do use WeakHashSet here, but WeakHashSet does not have amortizedCleanupIfNeeded(), like WeakHashMap does.

Maybe we should add amortizedCleanupIfNeeded() to WeakHashSet.

geoffreygaren reviewed (approved):
r=me

achristensen07 commented on Source/WebCore/Modules/screen-wake-lock/NavigatorScreenWakeLock.cpp:
```diff
@@ -61,7 +62,7 @@ WakeLock& NavigatorScreenWakeLock::wakeLock(Navigator& navigator)
 {
     if (!m_wakeLock)
-        m_wakeLock = WakeLock::create();
+        m_wakeLock = WakeLock::create(downcast<Document>(m_navigator.scriptExecutionContext()));
```
Why do we need to downcast this here?  It's just passed to the ContextDestructionObserver constructor.

cdumez replied:
I am purposefully using tight typing (Document instead of ScriptExecutionContext) to make it clear that WakeLock is only exposed to Window contexts and not workers.

I am not trying to future proof it to make it work with workers because I don't know that there is such plan. In particular, the spec doesn't currently expose this to workers. Also, if somebody were to add support for workers, there would likely be a lot more work involved than just changing the type here.

achristensen07 commented on Source/WebCore/Modules/screen-wake-lock/WakeLock.cpp:
```diff
@@ -26,16 +26,72 @@
+Document* WakeLock::document()
 {
-    promise->reject(Exception { NotSupportedError });
+    return downcast<Document>(scriptExecutionContext());
```
To be conservative, could we return null if it's not a Document?  (just in case someone in the future adds this to workers)

brentfulgham replied:
I think downcast<> does just return nullptr if the type is wrong?

cdumez replied:
But the type is not wrong. The WakeLock constructor takes a Document (because WakeLock is not exposed to workers) and thus its script execution context is always a Document.

achristensen07 commented on Source/WebCore/Modules/screen-wake-lock/WakeLockManager.h:
```diff
@@ -0,0 +1,58 @@
+private:
+    void visibilityStateChanged() final;
+
+    Document& m_document;
```
This is safe right now because the Document owns WakeLockManager with a std::unique_ptr.  We could make this a little more future-proof and make this a WeakPtr to protect against someone in the future making WakeLockManager RefCounted.

cdumez replied:
I hesitated. The reason I went this way is that it makes it clear the document cannot be null. With WeakPtr<document>, it is tempting to null check it everywhere.

brentfulgham commented on Source/WebCore/Modules/screen-wake-lock/WakeLock.cpp:
```diff
@@ -26,16 +26,72 @@
 
-WakeLock::WakeLock() = default;
+WakeLock::WakeLock(Document* document)
+    : ContextDestructionObserver(document)
```
It's funny that this paradigm seems to support the concept of being a context destruction observer for a nullptr.

brentfulgham commented on Source/WebCore/Modules/screen-wake-lock/WakeLockManager.cpp:
```diff
@@ -0,0 +1,105 @@
+        return;
+
+    switch (type) {
+    case WakeLockType::Screen:
```
It seems silly and inefficient to have a WakeLockType enum with a single value. I realize that this is how the specification is defined, but it is strange! Why pay the cost of a switch/case/branch here when it can never be anything else?

cdumez replied:
I wrote it this way to match the spec and for future extensibility as different types of lock can be supported (I think Blink has a system lock for e.g.). Hopefully the compiler is smart enough to optimize this out but if it isn't, this is not hot code at all.

cdumez committed:
[WK2] Add initial implementation the Screen Wake Lock API behind an experimental flag
https://bugs.webkit.org/show_bug.cgi?id=245712

Reviewed by Alex Christensen, Brent Fulgham and Geoffrey Garen.

Add initial implementation the Screen Wake Lock API for WebKit2, behind an experimental flag:
- https://w3c.github.io/screen-wake-lock

The implementation should be complete but it is currently behind an experimental feature flag
that is off by default.

Some tests are still failing. The reason for that seems to be that our support for Feature
Policy is only partial. In particular, we seem to support the `allow` attribute on iframes
but not the Feature-Policy HTTP header.

We also have to make a decision about whether or not we want to prompt the user when the
Website tries to take a screen wake lock and if we need an indicator in the browser UI.
Currently, my implementation only requires a user gesture.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/feature-policy/resources/feature-policy-screen-wakelock.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/feature-policy/resources/featurepolicy.js:
(expect_reports):
* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
(window.test_driver_internal.set_permission):
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/idlharness.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-active-document.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-disabled-by-feature-policy.https.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-disabled-by-feature-policy.https.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-by-feature-policy-attribute-redirect-on-load.https.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-by-feature-policy-attribute-redirect-on-load.https.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-by-feature-policy-attribute.https.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-by-feature-policy-attribute.https.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-by-feature-policy.https.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-by-feature-policy.https.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-on-self-origin-by-feature-policy.https.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-enabled-on-self-origin-by-feature-policy.https.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-onrelease.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-released.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-request-denied.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelock-type.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-wake-lock/wakelockpermissiondescriptor.https-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/Modules/permissions/PermissionName.h:
* Source/WebCore/Modules/permissions/PermissionName.idl:
* Source/WebCore/Modules/screen-wake-lock/NavigatorScreenWakeLock.cpp:
(WebCore::NavigatorScreenWakeLock::NavigatorScreenWakeLock):
(WebCore::NavigatorScreenWakeLock::wakeLock):
* Source/WebCore/Modules/screen-wake-lock/NavigatorScreenWakeLock.h:
* Source/WebCore/Modules/screen-wake-lock/WakeLock.cpp:
(WebCore::WakeLock::WakeLock):
(WebCore::WakeLock::request):
(WebCore::WakeLock::document):
* Source/WebCore/Modules/screen-wake-lock/WakeLock.h:
(WebCore::WakeLock::create):
* Source/WebCore/Modules/screen-wake-lock/WakeLockManager.cpp: Added.
(WebCore::WakeLockManager::WakeLockManager):
(WebCore::WakeLockManager::~WakeLockManager):
(WebCore::WakeLockManager::addWakeLock):
(WebCore::WakeLockManager::removeWakeLock):
(WebCore::WakeLockManager::visibilityStateChanged):
(WebCore::WakeLockManager::releaseAllLocks):
* Source/WebCore/Modules/screen-wake-lock/WakeLockManager.h: Copied from Source/WebCore/Modules/screen-wake-lock/WakeLock.h.
* Source/WebCore/Modules/screen-wake-lock/WakeLockSentinel.cpp:
(WebCore::WakeLockSentinel::release):
(WebCore::WakeLockSentinel::virtualHasPendingActivity const):
(WebCore::WakeLockSentinel::eventListenersDidChange):
(WebCore::WakeLockSentinel::wakeLockManager):
* Source/WebCore/Modules/screen-wake-lock/WakeLockSentinel.h:
* Source/WebCore/Modules/screen-wake-lock/WakeLockType.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::wakeLockManager):
(WebCore::Document::stopActiveDOMObjects):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/TaskSource.h:
* Source/WebCore/html/FeaturePolicy.cpp:
(WebCore::policyTypeName):
(WebCore::FeaturePolicy::parse):
(WebCore::FeaturePolicy::allows const):
* Source/WebCore/html/FeaturePolicy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::queryPermission):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::setScreenWakeLockPermission):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.h:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setScreenWakeLockPermission):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::handleQueryPermission):
(WTR::TestController::resetStateToConsistentValues):
(WTR::TestController::setScreenWakeLockPermission):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):

Canonical link: https://commits.webkit.org/255019@main
